rpc/client/httpclient.go

```diff
@@ -18,67 +18,170 @@ import (
 )
 
 /*
-HTTP is a Client implementation that communicates with a tendermint node over
-json rpc and websockets.
+HTTP is a Client implementation that communicates with a Tendermint node over
+JSON RPC and WebSockets.
 
 This is the main implementation you probably want to use in production code.
-There are other implementations when calling the tendermint node in-process
+There are other implementations when calling the Tendermint node in-process
 (Local), or when you want to mock out the server for test code (mock).
 
 You can subscribe for any event published by Tendermint using Subscribe method.
-Note delivery is best-effort. If you don't read events fast enough or network
-is slow, Tendermint might cancel the subscription. The client will attempt to
+Note delivery is best-effort. If you don't read events fast enough or network is
+slow, Tendermint might cancel the subscription. The client will attempt to
 resubscribe (you don't need to do anything). It will keep trying every second
 indefinitely until successful.
+
+Request batching is available for JSON RPC requests over HTTP, which conforms to
+the JSON RPC specification (https://www.jsonrpc.org/specification#batch). See
+the example for more details.
 */
 type HTTP struct {
 	remote string
 	rpc    *rpcclient.JSONRPCClient
+
+	*baseRPCClient
 	*WSEvents
 }
 
-// NewHTTP takes a remote endpoint in the form tcp://<host>:<port>
-// and the websocket path (which always seems to be "/websocket")
+// BatchHTTP provides the same interface as `HTTP`, but allows for batching of
+// requests (as per https://www.jsonrpc.org/specification#batch). Do not
+// instantiate directly - rather use the HTTP.NewBatch() method to create an
+// instance of this struct.
+//
+// Batching of HTTP requests is thread-safe in the sense that multiple
+// goroutines can each create their own batches and send them using the same
+// HTTP client. Multiple goroutines could also enqueue transactions in a single
+// batch, but ordering of transactions in the batch cannot be guaranteed in such
+// an example.
+type BatchHTTP struct {
+	rpcBatch *rpcclient.JSONRPCRequestBatch
+	*baseRPCClient
+}
+
+// rpcClient is an internal interface to which our RPC clients (batch and
+// non-batch) must conform. Acts as an additional code-level sanity check to
+// make sure the implementations stay coherent.
+type rpcClient interface {
```

Comment:
Why not use already available public interface?

Reply:
Yeah, that is a good point 👍  I think it is OK to have this private `rpcClient` interface but it should make use of the existing interfaces by embedding them here.

Reply:
I can make `rpcClient` like this:

```go
type rpcClient interface {
	ABCIClient
	HistoryClient
	NetworkClient
	SignClient
	StatusClient
}
```

But I'm trying to avoid including the `EventsClient` interface in the `BatchHTTP` struct.